File: libs/moqtail-ts/src/model/extension_header/object_extension.ts
/**
 * Copyright 2026 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { KeyValuePair } from '../common/pair'
import { ProtocolViolationError } from '../error/error'
import { TrackExtensionType } from './constant'

export class ImmutableExtensionsObjectExtension {
  static readonly TYPE = TrackExtensionType.ImmutableExtensions

  constructor(public readonly extensions: KeyValuePair[]) {}

  toKeyValuePair(): KeyValuePair {
    const inner = new ByteBuffer()
    for (const kvp of this.extensions) {
      inner.putBytes(kvp.serialize().toUint8Array())
    }
    return KeyValuePair.tryNewBytes(ImmutableExtensionsObjectExtension.TYPE, inner.toUint8Array())
  }

  static fromKeyValuePair(pair: KeyValuePair): ImmutableExtensionsObjectExtension | undefined {
    if (Number(pair.typeValue) !== ImmutableExtensionsObjectExtension.TYPE || !(pair.value instanceof Uint8Array))
      return undefined
    const buf = new FrozenByteBuffer(pair.value)
    const extensions: KeyValuePair[] = []
    while (buf.remaining > 0) {
      const inner = KeyValuePair.deserialize(buf)
      if (Number(inner.typeValue) === TrackExtensionType.ImmutableExtensions) {
        throw new ProtocolViolationError(
          'ImmutableExtensionsObjectExtension.fromKeyValuePair',
          'ImmutableExtensions must not contain nested ImmutableExtensions (0x0B)',
        )
      }
      extensions.push(inner)
    }
    return new ImmutableExtensionsObjectExtension(extensions)
  }
}

export class PriorGroupIdGapExtension {
  static readonly TYPE = TrackExtensionType.PriorGroupIdGap

  constructor(public readonly gap: bigint) {}

  toKeyValuePair(): KeyValuePair {
    return KeyValuePair.tryNewVarInt(PriorGroupIdGapExtension.TYPE, this.gap)
  }

  static fromKeyValuePair(pair: KeyValuePair): PriorGroupIdGapExtension | undefined {
    if (Number(pair.typeValue) !== PriorGroupIdGapExtension.TYPE || typeof pair.value !== 'bigint') return undefined
    return new PriorGroupIdGapExtension(pair.value)
  }
}

export class PriorObjectIdGapExtension {
  static readonly TYPE = TrackExtensionType.PriorObjectIdGap

  constructor(public readonly gap: bigint) {}

  toKeyValuePair(): KeyValuePair {
    return KeyValuePair.tryNewVarInt(PriorObjectIdGapExtension.TYPE, this.gap)
  }

  static fromKeyValuePair(pair: KeyValuePair): PriorObjectIdGapExtension | undefined {
    if (Number(pair.typeValue) !== PriorObjectIdGapExtension.TYPE || typeof pair.value !== 'bigint') return undefined
    return new PriorObjectIdGapExtension(pair.value)
  }
}

export class UnknownObjectExtension {
  constructor(public readonly kvp: KeyValuePair) {}

  toKeyValuePair(): KeyValuePair {
    return this.kvp
  }
}

export type ObjectExtension =
  | ImmutableExtensionsObjectExtension
  | PriorGroupIdGapExtension
  | PriorObjectIdGapExtension
  | UnknownObjectExtension

export namespace ObjectExtension {
  export function fromKeyValuePair(pair: KeyValuePair): ObjectExtension {
    return (
      ImmutableExtensionsObjectExtension.fromKeyValuePair(pair) ??
      PriorGroupIdGapExtension.fromKeyValuePair(pair) ??
      PriorObjectIdGapExtension.fromKeyValuePair(pair) ??
      new UnknownObjectExtension(pair)
    )
  }

  export function toKeyValuePair(ext: ObjectExtension): KeyValuePair {
    return ext.toKeyValuePair()
  }

  export function deserializeAll(buf: BaseByteBuffer): ObjectExtension[] {
    const result: ObjectExtension[] = []
    while (buf.remaining > 0) {
      const kvp = KeyValuePair.deserialize(buf)
      result.push(fromKeyValuePair(kvp))
    }
    return result
  }

  export function serializeInto(exts: ObjectExtension[], payload: ByteBuffer): void {
    for (const ext of exts) {
      payload.putBytes(ext.toKeyValuePair().serialize().toUint8Array())
    }
  }

  export function isImmutableExtensions(ext: ObjectExtension): ext is ImmutableExtensionsObjectExtension {
    return ext instanceof ImmutableExtensionsObjectExtension
  }

  export function isPriorGroupIdGap(ext: ObjectExtension): ext is PriorGroupIdGapExtension {
    return ext instanceof PriorGroupIdGapExtension
  }

  export function isPriorObjectIdGap(ext: ObjectExtension): ext is PriorObjectIdGapExtension {
    return ext instanceof PriorObjectIdGapExtension
  }

  export function isUnknown(ext: ObjectExtension): ext is UnknownObjectExtension {
    return ext instanceof UnknownObjectExtension
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  function roundtrip(exts: ObjectExtension[]): ObjectExtension[] {
    const buf = new ByteBuffer()
    ObjectExtension.serializeInto(exts, buf)
    return ObjectExtension.deserializeAll(buf.freeze())
  }

  describe('ObjectExtension', () => {
    test('empty deserializeAll returns []', () => {
      const buf = new FrozenByteBuffer(new Uint8Array())
      expect(ObjectExtension.deserializeAll(buf)).toEqual([])
    })

    test('PriorGroupIdGapExtension roundtrip', () => {
      const ext = new PriorGroupIdGapExtension(7n)
      const [result] = roundtrip([ext])
      expect(result).toBeInstanceOf(PriorGroupIdGapExtension)
      expect((result as PriorGroupIdGapExtension).gap).toBe(7n)
    })

    test('PriorObjectIdGapExtension roundtrip', () => {
      const ext = new PriorObjectIdGapExtension(3n)
      const [result] = roundtrip([ext])
      expect(result).toBeInstanceOf(PriorObjectIdGapExtension)
      expect((result as PriorObjectIdGapExtension).gap).toBe(3n)
    })

    test('ImmutableExtensionsObjectExtension roundtrip', () => {
      const inner = [KeyValuePair.tryNewVarInt(0x04, 100n)]
      const ext = new ImmutableExtensionsObjectExtension(inner)
      const [result] = roundtrip([ext])
      expect(result).toBeInstanceOf(ImmutableExtensionsObjectExtension)
      expect((result as ImmutableExtensionsObjectExtension).extensions.length).toBe(1)
    })

    test('ImmutableExtensionsObjectExtension throws on nested 0x0B', () => {
      const nested = new ImmutableExtensionsObjectExtension([])
      const innerKvp = nested.toKeyValuePair()
      const outer = new ByteBuffer()
      outer.putBytes(innerKvp.serialize().toUint8Array())
      const outerKvp = KeyValuePair.tryNewBytes(TrackExtensionType.ImmutableExtensions, outer.toUint8Array())
      expect(() => ImmutableExtensionsObjectExtension.fromKeyValuePair(outerKvp)).toThrow(ProtocolViolationError)
    })

    test('UnknownObjectExtension pass-through roundtrip', () => {
      const kvp = KeyValuePair.tryNewVarInt(0x02, 42n)
      const ext = new UnknownObjectExtension(kvp)
      const [result] = roundtrip([ext])
      expect(result).toBeInstanceOf(UnknownObjectExtension)
      expect((result as UnknownObjectExtension).kvp.typeValue).toBe(0x02n)
    })

    test('mixed list roundtrip', () => {
      const exts: ObjectExtension[] = [
        new PriorGroupIdGapExtension(5n),
        new PriorObjectIdGapExtension(2n),
        new ImmutableExtensionsObjectExtension([]),
      ]
      const result = roundtrip(exts)
      expect(result.length).toBe(3)
      expect(result[0]).toBeInstanceOf(PriorGroupIdGapExtension)
      expect(result[1]).toBeInstanceOf(PriorObjectIdGapExtension)
      expect(result[2]).toBeInstanceOf(ImmutableExtensionsObjectExtension)
    })
  })
}

File: libs/moqtail-rs/src/model/extension_header/object_extension.rs
// Copyright 2026 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::constant::TrackExtensionType;
use crate::model::common::pair::KeyValuePair;
use crate::model::error::ParseError;
use bytes::{Buf, Bytes, BytesMut};

/// Typed representation of a MOQT object-level Extension Header.
///
/// Unknown extension types are preserved as raw KeyValuePairs so that relays
/// can forward them unchanged per spec.
#[derive(Debug, Clone, PartialEq)]
pub enum ObjectExtension {
  ImmutableExtensions { extensions: Vec<KeyValuePair> },
  PriorGroupIdGap { gap: u64 },
  PriorObjectIdGap { gap: u64 },
  Unknown { kvp: KeyValuePair },
}

impl ObjectExtension {
  /// Returns the raw wire type value for this extension.
  pub fn type_value(&self) -> u64 {
    match self {
      Self::ImmutableExtensions { .. } => TrackExtensionType::ImmutableExtensions as u64,
      Self::PriorGroupIdGap { .. } => TrackExtensionType::PriorGroupIdGap as u64,
      Self::PriorObjectIdGap { .. } => TrackExtensionType::PriorObjectIdGap as u64,
      Self::Unknown { kvp } => kvp.get_type(),
    }
  }

  /// Serializes this extension to its wire KVP bytes.
  pub fn serialize(&self) -> Result<Bytes, ParseError> {
    let kvp: KeyValuePair = self.clone().try_into()?;
    kvp.serialize()
  }

  /// Deserializes an ObjectExtension from a pre-parsed KeyValuePair.
  /// Unknown type values are returned as `Unknown { kvp }` rather than an error.
  pub fn deserialize(kvp: KeyValuePair) -> Result<Self, ParseError> {
    let type_value = kvp.get_type();
    let ext_type = match TrackExtensionType::try_from(type_value) {
      Ok(t) => t,
      Err(_) => return Ok(Self::Unknown { kvp }),
    };

    match ext_type {
      TrackExtensionType::ImmutableExtensions => {
        let bytes = kvp_bytes_value(&kvp, "ObjectExtension::deserialize(ImmutableExtensions)")?;
        let mut buf = bytes;
        let mut extensions = Vec::new();
        while buf.has_remaining() {
          let inner = KeyValuePair::deserialize(&mut buf)?;
          if inner.get_type() == TrackExtensionType::ImmutableExtensions as u64 {
            return Err(ParseError::ProtocolViolation {
              context: "ObjectExtension::deserialize(ImmutableExtensions)",
              details: "ImmutableExtensions MUST NOT contain another ImmutableExtensions key"
                .to_string(),
            });
          }
          extensions.push(inner);
        }
        Ok(Self::ImmutableExtensions { extensions })
      }
      TrackExtensionType::PriorGroupIdGap => {
        let value = kvp_varint_value(&kvp, "ObjectExtension::deserialize(PriorGroupIdGap)")?;
        Ok(Self::PriorGroupIdGap { gap: value })
      }
      TrackExtensionType::PriorObjectIdGap => {
        let value = kvp_varint_value(&kvp, "ObjectExtension::deserialize(PriorObjectIdGap)")?;
        Ok(Self::PriorObjectIdGap { gap: value })
      }
      // Track-scope extensions; treated as unknown at object level
      _ => Ok(Self::Unknown { kvp }),
    }
  }
}

impl TryInto<KeyValuePair> for ObjectExtension {
  type Error = ParseError;

  fn try_into(self) -> Result<KeyValuePair, Self::Error> {
    match self {
      Self::ImmutableExtensions { extensions } => {
        let mut buf = BytesMut::new();
        for ext in &extensions {
          buf.extend_from_slice(&ext.serialize()?);
        }
        KeyValuePair::try_new_bytes(TrackExtensionType::ImmutableExtensions as u64, buf.freeze())
      }
      Self::PriorGroupIdGap { gap } => {
        KeyValuePair::try_new_varint(TrackExtensionType::PriorGroupIdGap as u64, gap)
      }
      Self::PriorObjectIdGap { gap } => {
        KeyValuePair::try_new_varint(TrackExtensionType::PriorObjectIdGap as u64, gap)
      }
      Self::Unknown { kvp } => Ok(kvp),
    }
  }
}

/// Serializes a slice of ObjectExtensions to their concatenated wire KVP bytes.
pub fn serialize_object_extensions(exts: &[ObjectExtension]) -> Result<Bytes, ParseError> {
  let mut buf = BytesMut::new();
  for ext in exts {
    buf.extend_from_slice(&ext.serialize()?);
  }
  Ok(buf.freeze())
}

/// Deserializes ObjectExtensions from a byte slice of known length.
/// Reads KVPs until the buffer is empty.
pub fn deserialize_object_extensions(
  bytes: &mut Bytes,
) -> Result<Vec<ObjectExtension>, ParseError> {
  let mut extensions = Vec::new();
  while bytes.has_remaining() {
    let kvp = KeyValuePair::deserialize(bytes)?;
    extensions.push(ObjectExtension::deserialize(kvp)?);
  }
  Ok(extensions)
}

fn kvp_varint_value(kvp: &KeyValuePair, context: &'static str) -> Result<u64, ParseError> {
  match kvp {
    KeyValuePair::VarInt { value, .. } => Ok(*value),
    KeyValuePair::Bytes { type_value, .. } => Err(ParseError::ProtocolViolation {
      context,
      details: format!(
        "Extension type 0x{type_value:02X} expects VarInt encoding but received Bytes"
      ),
    }),
  }
}

fn kvp_bytes_value(kvp: &KeyValuePair, context: &'static str) -> Result<Bytes, ParseError> {
  match kvp {
    KeyValuePair::Bytes { value, .. } => Ok(value.clone()),
    KeyValuePair::VarInt { type_value, .. } => Err(ParseError::ProtocolViolation {
      context,
      details: format!(
        "Extension type 0x{type_value:02X} expects Bytes encoding but received VarInt"
      ),
    }),
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use bytes::Bytes;

  fn roundtrip(ext: ObjectExtension) -> ObjectExtension {
    let serialized = ext.serialize().unwrap();
    let mut buf = serialized;
    let kvp = KeyValuePair::deserialize(&mut buf).unwrap();
    assert!(!buf.has_remaining());
    ObjectExtension::deserialize(kvp).unwrap()
  }

  #[test]
  fn test_roundtrip_immutable_extensions() {
    let inner = vec![
      KeyValuePair::try_new_varint(0x10, 7).unwrap(),
      KeyValuePair::try_new_bytes(0x11, Bytes::from_static(b"data")).unwrap(),
    ];
    let ext = ObjectExtension::ImmutableExtensions {
      extensions: inner.clone(),
    };
    assert_eq!(roundtrip(ext.clone()), ext);
  }

  #[test]
  fn test_roundtrip_prior_group_id_gap() {
    let ext = ObjectExtension::PriorGroupIdGap { gap: 3 };
    assert_eq!(roundtrip(ext.clone()), ext);
  }

  #[test]
  fn test_roundtrip_prior_object_id_gap() {
    let ext = ObjectExtension::PriorObjectIdGap { gap: 5 };
    assert_eq!(roundtrip(ext.clone()), ext);
  }

  #[test]
  fn test_roundtrip_unknown() {
    let kvp = KeyValuePair::try_new_varint(0xFE, 42).unwrap();
    let ext = ObjectExtension::Unknown { kvp };
    assert_eq!(roundtrip(ext.clone()), ext);
  }

  #[test]
  fn test_immutable_extensions_nested_is_error() {
    let inner_immutable = KeyValuePair::try_new_bytes(0x0B, Bytes::from_static(b"")).unwrap();
    let inner_bytes = inner_immutable.serialize().unwrap();
    let outer = KeyValuePair::try_new_bytes(0x0B, inner_bytes).unwrap();
    assert!(matches!(
      ObjectExtension::deserialize(outer).unwrap_err(),
      ParseError::ProtocolViolation { .. }
    ));
  }

  #[test]
  fn test_serialize_deserialize_mixed() {
    let exts = vec![
      ObjectExtension::PriorGroupIdGap { gap: 2 },
      ObjectExtension::PriorObjectIdGap { gap: 1 },
      ObjectExtension::Unknown {
        kvp: KeyValuePair::try_new_varint(0xFC, 0).unwrap(),
      },
    ];
    let serialized = serialize_object_extensions(&exts).unwrap();
    let mut buf = serialized;
    let deserialized = deserialize_object_extensions(&mut buf).unwrap();
    assert_eq!(deserialized, exts);
  }

  #[test]
  fn test_serialize_deserialize_empty() {
    let serialized = serialize_object_extensions(&[]).unwrap();
    assert!(serialized.is_empty());
    let mut buf = serialized;
    let deserialized = deserialize_object_extensions(&mut buf).unwrap();
    assert!(deserialized.is_empty());
  }
}
